[LC-582] Set block header attributes according to P-Rep list status.

metalg0su opened the pull request:
# Goal
Set block header attributes according to P-Rep list status.

# Detail
Suppose that 3 cases when build & invoke block.
Each of cases has different block header attributes. Explain below.

## In Case Normal
No changes in P-Rep list

- next_leader: No changes (TODO: ignored when leader-complained?)
- reps_hash: No changes
- next_reps_hash: Set empty hash

## In Case Term End (Reset order)
P-Rep term expired and switched to new list

- next_leader: Set empty hash
- reps_hash: No changes
- next_reps_hash: Invoked value of `next prep root hash` from score

## In case Panelty (Keep Order)
One of P-Rep has been punished and excluded from list 

- next_leader: No changes
- reps_hash: No changes
- next_reps_hash: Invoked value of `next prep root hash` from score

metalg0su committed:
[LC-582] Set `next_leader` in score invoked block

- If P-Rep has been changed, there are two possible reasons:
 1. P-Rep term reaches to the end, and changed to new P-Reps.
 > Set next_leader as Hash32.empty()
 2. One of P-Rep is punished (Panelty).
 > Set next_leader as it is

- If P-Rep has no changes:
 > Set next_leader as it is

metalg0su committed:
[LC-582] Set `next_reps_hash` in score invoked block

metalg0su committed:
[LC-582] Add P-Rep changed reason

windies21 committed:
[LC-582] add find_preps_addresses_by_header to blockchain and remove get
rep ids.

windies21 committed:
[LC-582] add property preg_changed and revealed_next_reps_hash to block

windies21 committed:
[LC-582] Check new term with header.prep_changed.

metalg0su commented on loopchain/blockchain/blocks/v0_3/block.py:
```diff
@@ -25,6 +27,26 @@ class BlockHeader(BaseBlockHeader):
+        if self.prep_changed:
+            return self.next_reps_hash
+        return self.reps_hash
+
```
those three method became used at both v0.1 and v0.3.
How about extract those things at `BaseBlockHeader`?

windies21 replied:
good idea!

windies21 replied:
But it's not simple. Because BaseBlockHeader is set to the data class. I postpone this to the next one.

metalg0su commented on loopchain/blockchain/blocks/v0_3/block.py:
```diff
@@ -25,6 +27,26 @@ class BlockHeader(BaseBlockHeader):
         return self.leader_votes_hash != Hash32.empty()
 
+    @property
+    def prep_changed(self) -> Optional[PrepChangedReason]:
```
I think its naming could raise confusion to return bool type (prep changed or not).
How about `prep_changed_reason` or...

as normal method: get_prep_changed_reason? I saw in your code below:
```python
            if self._block_manager.epoch.leader_id != ChannelProperty().peer_id \
                    and not candidate_block.header.prep_changed:
```
It could be:
```python
            if self._block_manager.epoch.leader_id != ChannelProperty().peer_id \
                    and not candidate_block.header.get_prep_changed_reason():
```

windies21 replied:
It will develop further to feature 'keep order'. In this PR it was used as a kind of boolean (of course It's not actual bool value).

metalg0su commented on loopchain/blockchain/blocks/v0_3/block.py:
```diff
@@ -25,6 +27,26 @@ class BlockHeader(BaseBlockHeader):
+        return PrepChangedReason.PENALTY
+
+    @property
+    def revealed_next_reps_hash(self):
```
Can you explain `revealed` means?

metalg0su commented on loopchain/blockchain/blocks/v0_3/block_verifier.py:
```diff
@@ -113,14 +115,22 @@ def verify_invoke(self, builder: 'BlockBuilder', block: 'Block', prev_block: 'Bl
+                    and header.next_reps_hash == new_block.header.revealed_next_reps_hash:
+                pass
+            else:
+                exception = RuntimeError(f"Block({header.height}, {header.hash.hex()}, "
```
I think it is better to specify Exception Type for better legibility, like `PrepChangedButNoChangedInRepsHash` (--;)

windies21 replied:
Not interested in this PR

windies21 committed:
[LC-582] modify lru_chache caching condition.

yakkle reviewed (commented):
It would be better to read that `prep_changed` compare statements change like below.
```py
        if last_block.header.prep_changed == PrepChangedReason.PENALTY:
```

windies21 commented on loopchain/baseservice/score_code.py:
```diff
@@ -28,4 +28,4 @@ class ScoreResponse(IntEnum):
 class PrepChangedReason:
     TERM_END = "0x0"  # Term ends and reset P-Rep list
-    PANELTY = "0x1"  # Updated within term (Panelty)
+    PENALTY = "0x1"  # Updated within term (Penalty)
```
You can consider it for feature 'keep order' (LC-557). But It's not necessary in this PR.

windies21 commented on loopchain/blockchain/blocks/v0_3/block.py:
```diff
@@ -25,6 +27,26 @@ class BlockHeader(BaseBlockHeader):
+
+    @property
+    def revealed_next_reps_hash(self):
+        if self.prep_changed:
```
Nope. Normal block has empty hash in its next_reps_hash but That's not true value. That means this block is normal block. So Engine can revealed next reps hash from that block. That block's real next_reps_hash is same to reps_hash.

windies21 replied:
We have to check if this block is normal. Penalty check is not good enough.

yakkle replied:
When I see this line, I checked `prep_changed` function.
and check return value `None` or `PrepChangedReason.TERM_END` or `PrepChangedReason.PENALTY`.
Then, I have seen `PrepChangedReason` class to check real value of `TERM_END` and `PENALTY`.
at last, I got it `prep_changed` is True when `TERM_END`, `PENALTY`.

`prep_changed` function is not clear to understand `prep_changed` condition.
It disturb code reading flow.

So, I suggest like this for readability all over the `prep_changed` used.
```py
if self.prep_changed in (PrepChangedReason.TERM_END, PrepChangedReason.PENALTY):
```
or, change `prep_changed` function to `prep_changed_reason`.
and add `prep_changed` function.
```py
    def prep_changed(self) -> bool:
        return self.prep_changed_reason in (PrepChangedReason.TERM_END, PrepChangedReason.PENALTY):
```

yakkle commented on loopchain/peer/consensus_siever.py:
```diff
@@ -126,8 +126,8 @@ def __get_complaint_votes(self):
         async with self.__lock:
             if self._block_manager.epoch.leader_id != ChannelProperty().peer_id:
-                util.logger.warning(f"This peer is not leader. epoch leader={self._block_manager.epoch.leader_id}")
-                return
```
why this return line removed?

windies21 replied:
When the term transitions, the leader is not the leader but must make the last block. When leader make last block of Term, that leader is no more leader but must make the last block.

windies21 commented on loopchain/blockchain/blockchain.py:
```diff
@@ -1139,9 +1137,9 @@ def score_invoke(self, _block: Block, prev_block: Block, is_block_editable: bool
-            elif next_prep.state == PrepChangedReason.PANELTY:
+            if next_prep["state"] == PrepChangedReason.TERM_END:
+                next_leader = ExternalAddress.empty()
+            elif next_prep["state"] == PrepChangedReason.PENALTY:
```
You can consider it for feature 'keep order' (LC-557). But It's not necessary in this PR.

metalg0su committed:
[LC-582] Add unit test

yakkle commented on loopchain/blockchain/blocks/v0_3/block_verifier.py:
```diff
@@ -113,14 +115,22 @@ def verify_invoke(self, builder: 'BlockBuilder', block: 'Block', prev_block: 'Bl
+            if not new_block.header.prep_changed \
+                    and header.next_reps_hash == new_block.header.revealed_next_reps_hash:
+                pass
+            else:
```
I think `if (condition): else` statements can be replaced by the following condition.
```suggestion
            if (new_block.header.prep_changed or header.next_reps_hash != new_block.header.revealed_next_reps_hash):
```

windies21 replied:
Good suggestion.

windies21 committed:
[LC-582] add ConfirmInfoInvalidNeedBlockSync exception in verify_confirm_info

- add state 'BlockGenerate' in announce_unconfirmed_block to stop
consensus loop when there is excepts ConfirmInfoInvalidNeedBlockSync.

- Avoid error message when add_complain if there is no epoch.

- raise ConfirmInfoInvalidNeedBlockSync when receive higher block after reps change.

windies21 committed:
[LC-582] change condition in verify_invoke of block_verifier.

yakkle commented:
retest this please
